README.md
# Let-it-snow
## Synopsis

This code implements the snow cover extent detection algorithm LIS (Let It Snow) for Sentinel-2, Landsat-8 and SPOT4-Take5 data.

The algorithm documentation with examples is available here:

* [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1414452.svg)](https://doi.org/10.5281/zenodo.1414452)

Access to Theia Snow data collection:

* [![DOI:10.24400/329360/f7q52mnk](https://zenodo.org/badge/DOI/10.24400/329360/f7q52mnk.svg)](http://doi.org/10.24400/329360/f7q52mnk)

To read more about the "Centre d'Expertise Scientifique surface enneigée" (in French):

* [Bulletin THEIA](https://www.theia-land.fr/sites/default/files/imce/BulletinTHEIA3.pdf#page=10)

The input files are Sentinel-2 or Landsat-8 level-2A products from the [Theia Land Data Centre](https://theia.cnes.fr/) or [SPOT-4/5 Take 5 level-2A products](https://spot-take5.org) and a Digital Terrain Model (DTM) like SRTM for instance reprojected at the same resolution as the input image.

## Usage

Run the python script run_snow_detector.py with a json configuration file as unique argument:

```bash
python run_snow_detector.py param.json
```
The snow detection is performed in the Python script [run_snow_detector.py](app/run_snow_detector.py).

All the parameters of the algorithm, paths to input and output data are stored in the json file. See the provided example [param_test_s2_template.json](tes/param_test_s2_template.json) file for an example.

Moreover The JSON schema is available in the [Algorithm Theoretical Basis Documentation](doc/atbd/ATBD_CES-Neige.tex) and gives more information about the roles of these parameters.

NB: To build DEM data download the SRTM files corresponding to the study area and build the .vrt using gdalbuildvrt. Edit config.json file to activate preprocessing : Set "preprocessing" to true and set the vrt path.


## Products format

* COMPO: Raster image showing the outlines of the cloud (including cloud shadow) and snow masks drawn on the RGB composition of the L2A image (bands SWIR/Red/Green).
* SNOW_ALL: Binary mask of snow and clouds.
  * 1st bit: Snow mask after pass1
  * 2nd bit: Snow mask after pass2
  * 3rd bit: Clouds detected at pass0
  * 4th bit: Clouds refined  at pass0
  * 5th bit: Clouds initial (all_cloud)
  * 6th bit: Slope flag (optional 1: bad slope correction)

For example if you want to get the snow from pass1 and clouds detected from pass1 you need to do:
```python
pixel_value & 00000101
```
* SEB: Raster image of the snow mask and cloud mask.
  * 0: No-snow
  * 100: Snow
  * 205: Cloud including cloud shadow
  * 254: No data
* SEB_VEC: Vector image of the snow mask and cloud mask. Two fields of information are embedded in this product. DN (for Data Neige) and type.
  * DN field :
     * 0: No-snow
     * 100: Snow
     * 205: Cloud including cloud shadow
     * 254: No data

## Data set example

Sequence of snow maps produced from Sentinel-2 type of observations (SPOT-5 Take 5) over the Deux Alpes and Alpe d'Huez ski resorts are available on [Zenodo](http://doi.org/10.5281/zenodo.159563).

## Motivation

Code to generate the snow cover extent product on Theia platform.

## Installation

LIS processing chain uses CMake (http://www.cmake.org) for building from source.

### Dependencies

Following a summary of the required dependencies: 

* GDAL >=2.0
* OTB >= 6.2
* Python interpreter >= 2.7
* Python libs >= 2.7
* Python packages:
* numpy
* lxml
* matplotlib

GDAL itself depends on a number of other libraries provided by most major operating systems and also depends on the non standard GEOS and Proj libraries. GDAL- Python bindings are also required

Python package dependencies:

* sys
* subprocess
* glob
* os
* json
* gdal

Optional dependencies:

* gdal_trace_outline can be used alternatively to gdal_polygonize.py to generate the vector layer. It requires to install [dans-gdal-scripts utilities](https://github.com/gina-alaska/dans-gdal-scripts).

### Installing from the source distribution

#### General

In your build directory, use cmake to configure your build.
```bash
cmake -C config.cmake source_lis_path
```
In your config.cmake you need to set :
```bash
LIS_DATA_ROOT
```
For OTB superbuild users these cmake variables need to be set:
```bash
OTB_DIR
ITK_DIR
GDAL_INCLUDE_DIR
GDAL_LIBRARY
```
Run make in your build folder.
```bash
make
```
To install let-it-snow application and the s2snow python module.
In your build folder:
```bash
make install
```

Add appropriate executable rights
```bash
chmod -R 755 ${install_dir}
```

The files will be installed by default into /usr/local and add to the python default modules.
To overrsouride this behavior, the variable CMAKE_INSTALL_PREFIX must be configure before build step.

Update environment variables for LIS. Make sure that OTB and other dependencies directories are set in your environment variables:
```bash
export PATH=/your/install/directory/bin:/your/install/directory/app:$PATH
export LD_LIBRARY_PATH=/your/install/directory/lib:$LD_LIBRARY_PATH
export OTB_APPLICATION_PATH=/your/install/directory/lib:$OTB_APPLICATION_PATH
export PYTHONPATH=/your/install/directory/lib:/your/install/directory/lib/python2.7/site-packages:$PYTHONPATH
```
let-it-snow is now installed.

## Tests

Enable tests with BUILD_TESTING cmake option. Use ctest command to run tests. Do not forget to clean your output test directory when you run a new set of tests.

Data (input and baseline) to run validation tests are available on Zenodo:

* [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.166511.svg)](https://doi.org/10.5281/zenodo.166511)

Download LIS-Data and extract the folder. It contains all the data needed to run tests. Set Data-LIS path var in cmake configuration files.
Baseline : Baseline data folder. It contains output files of S2Snow that have been reviewed and validated.
Data-Test : Test data folder needed to run tests. It contains Landsat, Take5 and SRTM data.
Output-Test : Temporary output tests folder.
Do not modify these folders.

## Contributors

Manuel Grizonnet (CNES), Simon Gascoin (CNRS/CESBIO), Tristan Klempka (CNES), Germain Salgues (Magellium)

## License

This is free software under the GNU Affero General Public License v3.0. See
http://www.gnu.org/licenses/agpl.html for details.


CHANGELOG.md
# Change Log
All notable changes to Let It Snow (LIS) will be documented in this file.

## [1.5] - 2019-01-11

### Added
- The snow annual map module is now operational, the associated files are:
    - app/run_snow_annual_map.py, the main application
    - python/s2snow/snow_annual_map.py, the core of the annual map processing
    - python/s2snow/snow_annual_map_evaluation.py, provide the possibility to compare with other snow products and modis snow annual map
    - python/s2snow/snow_product_parser.py, class to handle the supported type of snow products
    - doc/snow_annual_map_schema.json, parameters descriptions
    - hpc/prepare_data_for_snow_annual_map.py, preprocessing script on CNES HPC, 
    - doc/tutorials/prepare_snow_annual_map_data.md, tutorial
- Provided new data pack for tests "Data-LIS-1.5"
- Add tests for snow annual map computation
- The version of the s2snow module is now stored in file python/s2snow/version.py
- Add support and tests for zipped products in build_json.py and run_snow_detector.py,
- Add a mode to build_json.py script to configure and run LIS on Level 2A MAJA native products
- Add a tutorial to describe how to run LIS on MAJA native products
- Add a mode to build_json.py script to configure and run LIS on SEN2COR Level 2A products
- Add a tutorial to describe how to run LIS on SEN2COR products
- Add a mode to build_json.py script to configure and run LIS on U.S. Landsat Analysis Ready Data (ARD) Level 2A products
- Add a tutorial to describe how to run LIS on Landsat ARD products	
- The expert mask now includes a 5th bit for the clouds that were present in the product/original cloud mask
- The expert mask now includes an optional 6th bit propagating the slope correction flag from the product mask when available
- The cold cloud removal (pass 1.5) now use an area threshold to process only significant snow areas within clouds and reduce time.
- Link ATBD and LIS Data for test validation to their Zenodo DOI in README.md

### Fixed
- Fix all python scripts headers to avoid to mix python versions
- Fix preprocessing json option which was broken which allows to resample input DTM
- Fix typos in README.md documentation
- Change nodata management (read -32768 from input and write 0 in the
    output) in DTM resampling to avoid error in snow line estimation
    on area without DTM information 

## [1.4] - 2018-02-14

### Added
- Experimental pass1_5 function implementing the removal of snow areas inside initial cloud mask (doughnuts)
- Experimental new application to run and evaluate an annual snow map computation from a timeserie of S2 and/or L8 snow products
- Added fclear_lim parameter minimum percentage of clear pixels in an elevation band
    (default value 0.1) used to compute the snow line.
- Added option to disable vector generation
- Added options to use and manage gdal_trace_outline instead of gdal_polygonize

### Changed
- Changed default value for parameter red_darkcloud to 300 to reduce cloud sensitivity
- Changed all_cloud_mask.tif, it now include the thin clouds (in accordance with ATBD)
- Fixed method compute_percent is fix when image is empty or filled with nodata
- Fixed zs condition to trigger properly pass2 (in accordance with ATBD)
- Changed zs computation in pass2 to considers the full image imprint (including nodata pixels)
- Changed cloud mask refinement is not apply during pass1 to improve snow line accuracy during pass2
- Updated build_json.py to handle boolean parameters
- Updated build_json.py to handle new lis input parameters

## [1.3.1] - 2017-11-23

### Fixed
- Fix the intermediate data format (used 1 bit instead of type uint8)

## [1.3] - 2017-11-02

### Added
- Use gdal_trace_outline from the gina-alaska package instead of gdal_polygonize if available

### Changed
- Move OTB minimum 6.0.0 which include a fix to handle properly 1 byte TIFF image
- Restore and document correct behaviour for LIS installation with install target(lib, bin,include, python) 
- New QGIS style files for raster and vector LIS product
- Use OTB Application Python API instead of call to subprocess
- Use Python Logging module for Python scripts instead of using print
- Changed compute_cloud_mask and compute_snow_mask by OTB applications
- Added a new app to generate the JSON configuration file (build_json.py)
- Changed the way the product is generated to avoid data duplication
- Change rasterize step to contour detection using 8 connectivity to generate the rgb composition
- Improved detection by adjusting default parameter red_pass2 from 0.120 to 0.40
- Improve code quality (pep8 and pylint)
- Improve installation instructions in the README.md
- Fix cpu usage to respect the "nb_threads" parameter set in the json file.
- The output product now use the input product directory name as PRODUCT_ID in the xml file.

## [1.2.1] - 2017-09-14
- Fix segfault in case number of histogram bins for the altitude channel is zero 

## [1.2] - 2017-06-04
- add json schema to ATBD to document all parameters
- Add version of lis in atbd
- Document how to build the documentation in doc/tex directory
- Compact histogram files and copy it in LIS_PRODUCTS
- Apply autopep8 to all Python scripts to improve code quality
- Add a changelog

## [1.1.1] - 2016-11-28
- minor update in build scripts
- change ctest launcher location

## [1.1.0] - 2016-11-28
- Change license from GPL to A-GPL
- Improvments in cmake configuration
- launch tests in separate directories

## [1.0.0] - 2016-07-06
- First released version of LIS with support with last MUSCATE format
- Support for image splitted in multiple files
- Use high clouds mask in cloud refine
- Prototype for multi-T synthesis


app/build_json.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import re
import sys
import json
import logging
import argparse
import zipfile

### Configuration Template ###
conf_template = {"general":{"pout":"",
                            "nodata":-10000,
                            "ram":1024,
                            "nb_threads":1,
                            "preprocessing":False,
                            "log":True,
                            "multi":1,
                            "target_resolution":-1},
                 "vector":{"generate_vector":True,
                           "generate_intermediate_vectors":False,
                           "use_gdal_trace_outline":True,
                           "gdal_trace_outline_dp_toler":0,
                           "gdal_trace_outline_min_area":0},
                 "inputs":{"green_band":{"path": "",
                                         "noBand": 1},
                           "red_band":{"path": "",
                                       "noBand": 1},
                           "swir_band":{"path": "",
                                        "noBand": 1},
                           "dem":"",
                           "cloud_mask":""},
                 "snow":{"dz":100,
                         "ndsi_pass1":0.4,
                         "red_pass1":200,
                         "ndsi_pass2":0.15,
                         "red_pass2":40,
                         "fsnow_lim":0.1,
                         "fclear_lim":0.1,
                         "fsnow_total_lim":0.001},
                 "cloud":{"shadow_in_mask":64,
                          "shadow_out_mask":128,
                          "all_cloud_mask":1,
                          "high_cloud_mask":32,
                          "rf":12,
                          "red_darkcloud":300,
                          "red_backtocloud":100,
                          "strict_cloud_mask":False,
                          "rm_snow_inside_cloud":False,
                          "rm_snow_inside_cloud_dilation_radius":1,
                          "rm_snow_inside_cloud_threshold":0.85,
                          "rm_snow_inside_cloud_min_area":5000}}


### Mission Specific Parameters ###

MAJA_parameters = {"multi":10,
                 "green_band":".*FRE_R1.DBL.TIF$",
                 "green_bandNumber":2,
                 "red_band":".*FRE_R1.DBL.TIF$",
                 "red_bandNumber":3,
                 "swir_band":".*FRE_R2.DBL.TIF$",
                 "swir_bandNumber":5,
                 "cloud_mask":".*CLD_R2.DBL.TIF$",
                 "dem":".*ALT_R2\.TIF$",
                 "shadow_in_mask":4,
                 "shadow_out_mask":8,
                 "all_cloud_mask":1,
                 "high_cloud_mask":128,
                 "rf":12}

SEN2COR_parameters = {"mode":"sen2cor",
                 "multi":10,
                 "green_band":".*_B03_10m.jp2$",
                 "green_bandNumber":1,
                 "red_band":".*_B04_10m.jp2$",
                 "red_bandNumber":1,
                 "swir_band":".*_B11_20m.jp2$",
                 "swir_bandNumber":1,
                 "cloud_mask":".*_SCL_20m.jp2$",
                 "dem":"",
                 "shadow_in_mask":3,
                 "shadow_out_mask":3,
                 "all_cloud_mask":8,
                 "high_cloud_mask":10,
                 "rf":12}

Take5_parameters = {"multi":1,
                    "green_band":".*ORTHO_SURF_CORR_PENTE.*\.TIF$",
                    "green_bandNumber":1,
                    "red_band":".*ORTHO_SURF_CORR_PENTE.*\.TIF$",
                    "red_bandNumber":2,
                    "swir_band":".*ORTHO_SURF_CORR_PENTE.*\.TIF$",
                    "swir_bandNumber":4,
                    "cloud_mask":".*NUA.*\.TIF$",
                    "div_mask":".*DIV.*\.TIF$",
                    "div_slope_thres":8,
                    "dem":".*\.tif",
                    "shadow_in_mask":64,
                    "shadow_out_mask":128,
                    "all_cloud_mask":1,
                    "high_cloud_mask":32,
                    "rf":8}

S2_parameters = {"multi":10,
                 "green_band":".*FRE_B3.*\.tif$",
                 "green_bandNumber":1,
                 "red_band":".*FRE_B4.*\.tif$",
                 "red_bandNumber":1,
                 "swir_band":".*FRE_B11.*\.tif$",
                 "swir_bandNumber":1,
                 "cloud_mask":".*CLM_R2.*\.tif$",
                 "dem":".*ALT_R2\.TIF$",
                 "div_mask":".*MG2_R2.*\.tif$",
                 "div_slope_thres":64,
                 "shadow_in_mask":32,
                 "shadow_out_mask":64,
                 "all_cloud_mask":1,
                 "high_cloud_mask":128,
                 "rf":12}

L8_parameters_new_format = {"multi":1,
                 "green_band":".*FRE_B3.*\.tif$",
                 "green_bandNumber":1,
                 "red_band":".*FRE_B4.*\.tif$",
                 "red_bandNumber":1,
                 "swir_band":".*FRE_B6.*\.tif$",
                 "swir_bandNumber":1,
                 "cloud_mask":".*CLM_XS.*\.tif$",
                 "dem":".*ALT_R2\.TIF$",
                 "div_mask":".*MG2_XS.*\.tif$",
                 "div_slope_thres":64,
                 "shadow_in_mask":32,
                 "shadow_out_mask":64,
                 "all_cloud_mask":1,
                 "high_cloud_mask":128,
                 "rf":8}

L8_parameters = {"multi":1,
                 "green_band":".*ORTHO_SURF_CORR_PENTE.*\.TIF$",
                 "green_bandNumber":3,
                 "red_band":".*ORTHO_SURF_CORR_PENTE.*\.TIF$",
                 "red_bandNumber":4,
                 "swir_band":".*ORTHO_SURF_CORR_PENTE.*\.TIF$",
                 "swir_bandNumber":6,
                 "cloud_mask":".*NUA.*\.TIF$",
                 "div_mask":".*DIV.*\.TIF$",
                 "div_slope_thres":8,
                 "dem":".*\.tif",
                 "shadow_in_mask":64,
                 "shadow_out_mask":128,
                 "all_cloud_mask":1,
                 "high_cloud_mask":32,
                 "rf":8}

LANDSAT8_LASRC_parameters = {"mode":"lasrc",
                 "multi":10,
                 "green_band":".*_sr_band3.tif$",
                 "green_bandNumber":1,
                 "red_band":".*_sr_band4.tif$",
                 "red_bandNumber":1,
                 "swir_band":".*_sr_band6.tif$",
                 "swir_bandNumber":1,
                 "cloud_mask":".*_pixel_qa.tif$",
                 "dem":".*\.tif",
                 "shadow_in_mask":8,
                 "shadow_out_mask":8,
                 "all_cloud_mask":224, # cloud with high confidence (32+(64+128)) 
                 "high_cloud_mask":800, # cloud and high cloud with high confidence (32 + (512+256)) 
                 "rf":8}

mission_parameters = {"S2":S2_parameters,\
                      "LANDSAT8":L8_parameters,\
                      "LANDSAT8_new_format":L8_parameters_new_format,\
                      "Take5":Take5_parameters,\
                      "MAJA":MAJA_parameters,\
                      "SEN2COR":SEN2COR_parameters,\
                      "LANDSAT8_LASRC":LANDSAT8_LASRC_parameters
                     }

def str2bool(v):
    if v.lower() in ('yes', 'true', 't', 'y', '1'):
        return True
    elif v.lower() in ('no', 'false', 'f', 'n', '0'):
        return False
    else:
        raise argparse.ArgumentTypeError('Boolean value expected.')

def findFiles(folder, pattern):
    """ Search recursively into a folder to find a patern match
    """
    matches = []
    if folder.lower().endswith('.zip'):
        zfile = zipfile.ZipFile(folder)
        for filename in zfile.namelist():
            if re.match(pattern, filename):
                matches.append("/vsizip/"+os.path.join(folder, filename))
    else:
        for root, dirs, files in os.walk(folder):
            for file in files:
                if re.match(pattern, file):
                    matches.append(os.path.join(root, file))
    return matches

def read_product(inputPath, mission):
    """ Read the content of the input product folder
    and load the data information required for snow detection.
    """
    if os.path.exists(inputPath):
        params = mission_parameters[mission]
        conf_json = conf_template

        conf_json["general"]["multi"] = params["multi"]
        conf_json["inputs"]["green_band"]["path"] = findFiles(inputPath, params["green_band"])[0]
        conf_json["inputs"]["red_band"]["path"] = findFiles(inputPath, params["red_band"])[0]
        conf_json["inputs"]["swir_band"]["path"] = findFiles(inputPath, params["swir_band"])[0]
        conf_json["inputs"]["green_band"]["noBand"] = params["green_bandNumber"]
        conf_json["inputs"]["red_band"]["noBand"] = params["red_bandNumber"]
        conf_json["inputs"]["swir_band"]["noBand"] = params["swir_bandNumber"]
        conf_json["inputs"]["cloud_mask"] = findFiles(inputPath, params["cloud_mask"])[0]
        result = findFiles(os.path.join(inputPath, "SRTM"), params["dem"])
        if result:
            conf_json["inputs"]["dem"] = result[0]
        else:
            logging.warning("No DEM found within product!")

        #  Check optional div mask parameters to access slope correction flags
        if "div_mask" in params and "div_slope_thres" in params:
            div_mask_tmp = findFiles(inputPath, params["div_mask"])
            if div_mask_tmp:
                conf_json["inputs"]["div_mask"] = div_mask_tmp[0]
                conf_json["inputs"]["div_slope_thres"] = params["div_slope_thres"]
            else:
                logging.warning("div_mask was not found, the slope correction flag will be ignored")

        conf_json["cloud"]["shadow_in_mask"] = params["shadow_in_mask"]
        conf_json["cloud"]["shadow_out_mask"] = params["shadow_out_mask"]
        conf_json["cloud"]["all_cloud_mask"] = params["all_cloud_mask"]
        conf_json["cloud"]["high_cloud_mask"] = params["high_cloud_mask"]
        conf_json["cloud"]["rf"] = params["rf"]

        #Check if an optional mode is provided in the mission configuration
        # Use in case of SEN2COR to handle differences between maja and sen2cor encoding
        if 'mode' in params:
            conf_json["general"]["mode"] = params["mode"]
 
        
        return conf_json
    else:
        logging.error(inputPath + " doesn't exist.")

def main():
    # Parse arguments
    parser = argparse.ArgumentParser(description='This script is used to \
                                generate the snow detector configuration json file.\
                                This configuration requires at least the input product path\
                                and the output path in which will be generated snow product.')

    parser.add_argument("inputPath", help="input product path \
                                         (supports S2/L8/Take5 products)")
    parser.add_argument("outputPath", help="output folder for the json configuration file, \
                                and also the configured output path for the snow product")

    group_general = parser.add_argument_group('general', 'general parameters')
    group_general.add_argument("-nodata", type=int)
    group_general.add_argument("-ram", type=int)
    group_general.add_argument("-nb_threads", type=int)
    group_general.add_argument("-generate_vector", type=str2bool, help="true/false")
    group_general.add_argument("-preprocessing", type=str2bool, help="true/false")
    group_general.add_argument("-log", type=str2bool, help="true/false")
    group_general.add_argument("-multi", type=float)
    group_general.add_argument("-target_resolution", type=float)


    group_inputs = parser.add_argument_group('inputs', 'input files')
    group_inputs.add_argument("-dem", help="dem file path, to use for processing the input product")
    group_inputs.add_argument("-cloud_mask", help="cloud mask file path")

    group_snow = parser.add_argument_group('snow', 'snow parameters')
    group_snow.add_argument("-dz", type=int)
    group_snow.add_argument("-ndsi_pass1", type=float)
    group_snow.add_argument("-red_pass1", type=float)
    group_snow.add_argument("-ndsi_pass2", type=float)
    group_snow.add_argument("-red_pass2", type=float)
    group_snow.add_argument("-fsnow_lim", type=float)
    group_snow.add_argument("-fsnow_total_lim", type=float)

    group_cloud = parser.add_argument_group('cloud', 'cloud parameters')
    group_cloud.add_argument("-shadow_in_mask", type=int)
    group_cloud.add_argument("-shadow_out_mask", type=int)
    group_cloud.add_argument("-all_cloud_mask", type=int)
    group_cloud.add_argument("-high_cloud_mask", type=int)
    group_cloud.add_argument("-rf", type=int)
    group_cloud.add_argument("-red_darkcloud", type=int)
    group_cloud.add_argument("-red_backtocloud", type=int)
    group_cloud.add_argument("-strict_cloud_mask", type=str2bool, help="true/false")

    args = parser.parse_args()

    inputPath = os.path.abspath(args.inputPath)
    outputPath = os.path.abspath(args.outputPath)

    sentinel2Acronyms = ['S2', 'SENTINEL2', 'S2A', 'S2B']
    
    # Test if it is a MAJA output products (generated with MAJA processor version XX)
    # FIXME: This detection based on directory substring detection is very week and error prone
    # FIXME: use a factory and detect by using xml metadata
    if '.SAFE' in inputPath:
        # L2A SEN2COR product
        logging.info("SEN2COR product detected (detect .SAFE in the input path...).")
        jsonData = read_product(inputPath, "SEN2COR")
    elif '.DBL.DIR' in inputPath:
        if any(s in inputPath for s in sentinel2Acronyms):
            logging.info("MAJA native product detected (detect .DBL.DIR substring in input path...)")
            jsonData = read_product(inputPath, "MAJA")
        else:
            logging.error("Only MAJA products from Sentinels are supported by build_json.py script for now.")
    elif any(s in inputPath for s in sentinel2Acronyms):
        logging.info("THEIA Sentinel product detected.")
        jsonData = read_product(inputPath, "S2")
    elif "Take5" in inputPath:
        logging.info("THEIA Take5 product detected.")
        jsonData = read_product(inputPath, "Take5")
    elif "LANDSAT8-OLITIRS-XS" in inputPath:
        logging.info("THEIA LANDSAT8 product detected. (new version)")
        jsonData = read_product(inputPath, "LANDSAT8_new_format")
    elif "LANDSAT8" in inputPath:
        logging.info("THEIA LANDSAT8 product detected.")
        jsonData = read_product(inputPath, "LANDSAT8")
    elif "LC08" in inputPath:
        logging.info("LANDSAT8 LASRC) product detected (LC08_L1TP in input path...).")
        jsonData = read_product(inputPath, "LANDSAT8_LASRC")
    else:
        logging.error("Unknown product type.")
        sys.exit(0)

    if jsonData:
        if not os.path.exists(outputPath):
            logging.info("Create directory " + outputPath + "...")
            os.makedirs(outputPath)

        jsonData["general"]["pout"] = outputPath

        # Override parameters for group general
        if args.nodata is not None:
            jsonData["general"]["nodata"] = args.nodata
        if args.preprocessing is not None:
            jsonData["general"]["preprocessing"] = args.preprocessing
        if args.generate_vector is not None:
            jsonData["vector"]["generate_vector"] = args.generate_vector
        if args.log is not None:
            jsonData["general"]["log"] = args.log
        if args.ram:
            jsonData["general"]["ram"] = args.ram
        if args.nb_threads:
            jsonData["general"]["nb_threads"] = args.nb_threads
        if args.multi:
            jsonData["general"]["multi"] = args.multi
        if args.target_resolution:
            jsonData["general"]["target_resolution"] = args.target_resolution

        # Override dem location
        if args.dem:
            jsonData["inputs"]["dem"] = os.path.abspath(args.dem)
            logging.warning("Using optional external DEM!")
        # Override cloud mask location
        if args.cloud_mask:
            jsonData["inputs"]["cloud_mask"] = os.path.abspath(args.cloud_mask)

        # Override parameters for group snow
        if args.dz:
            jsonData["snow"]["dz"] = args.dz
        if args.ndsi_pass1:
            jsonData["snow"]["ndsi_pass1"] = args.ndsi_pass1
        if args.red_pass1:
            jsonData["snow"]["red_pass1"] = args.red_pass1
        if args.ndsi_pass2:
            jsonData["snow"]["ndsi_pass2"] = args.ndsi_pass2
        if args.red_pass2:
            jsonData["snow"]["red_pass2"] = args.red_pass2
        if args.fsnow_lim:
            jsonData["snow"]["fsnow_lim"] = args.fsnow_lim
        if args.fsnow_total_lim:
            jsonData["snow"]["fsnow_total_lim"] = args.fsnow_total_lim

        # Override parameters for group cloud
        if args.shadow_in_mask:
            jsonData["cloud"]["shadow_in_mask"] = args.shadow_in_mask
        if args.shadow_out_mask:
            jsonData["cloud"]["shadow_out_mask"] = args.shadow_out_mask
        if args.all_cloud_mask:
            jsonData["cloud"]["all_cloud_mask"] = args.all_cloud_mask
        if args.high_cloud_mask:
            jsonData["cloud"]["high_cloud_mask"] = args.high_cloud_mask
        if args.rf:
            jsonData["cloud"]["rf"] = args.rf
        if args.red_darkcloud:
            jsonData["cloud"]["red_darkcloud"] = args.red_darkcloud
        if args.red_backtocloud:
            jsonData["cloud"]["red_backtocloud"] = args.red_backtocloud
        if args.strict_cloud_mask:
            jsonData["cloud"]["strict_cloud_mask"] = args.strict_cloud_mask

        if not jsonData["inputs"].get("dem"):
            logging.error("No DEM found!")
            return 1

        jsonFile = open(os.path.join(outputPath, "param_test.json"), "w")
        jsonFile.write(json.dumps(jsonData, indent=4))
        jsonFile.close()

if __name__ == "__main__":
    # Set logging level and format.
    logging.basicConfig(stream=sys.stdout, level=logging.INFO, format=\
        '%(asctime)s - %(filename)s:%(lineno)s - %(levelname)s - %(message)s')
    main()
